File: apps/flutter_parent/lib/screens/announcements/announcement_details_screen.dart
// Copyright (C) 2020 - present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import 'package:flutter/material.dart';
import 'package:flutter_parent/l10n/app_localizations.dart';
import 'package:flutter_parent/models/attachment.dart';
import 'package:flutter_parent/screens/announcements/announcement_view_state.dart';
import 'package:flutter_parent/utils/common_widgets/attachment_indicator_widget.dart';
import 'package:flutter_parent/utils/common_widgets/error_panda_widget.dart';
import 'package:flutter_parent/utils/common_widgets/loading_indicator.dart';
import 'package:flutter_parent/utils/common_widgets/web_view/html_description_tile.dart';
import 'package:flutter_parent/utils/core_extensions/date_time_extensions.dart';
import 'package:flutter_parent/utils/service_locator.dart';

import 'announcement_details_interactor.dart';

enum AnnouncementType { INSTITUTION, COURSE }

class AnnouncementDetailScreen extends StatefulWidget {
  final String announcementId;
  final String courseId;
  final AnnouncementType announcementType;

  AnnouncementDetailScreen(this.announcementId, this.announcementType, this.courseId, BuildContext? context, {super.key});

  @override
  State createState() => _AnnouncementDetailScreenState();
}

class _AnnouncementDetailScreenState extends State<AnnouncementDetailScreen> {
  Future<AnnouncementViewState?>? _announcementFuture;

  Future<AnnouncementViewState?> _loadAnnouncement(BuildContext context, {bool forceRefresh = false}) =>
      _interactor.getAnnouncement(
        widget.announcementId,
        widget.announcementType,
        widget.courseId,
        L10n(context).globalAnnouncementTitle,
        forceRefresh,
      );

  get _interactor => locator<AnnouncementDetailsInteractor>();

  @override
  Widget build(BuildContext context) {
    if (_announcementFuture == null) {
      _announcementFuture = _loadAnnouncement(context);
    }
    return FutureBuilder(
      future: _announcementFuture,
      builder: (context, AsyncSnapshot<AnnouncementViewState?> snapshot) {
        if (snapshot.connectionState == ConnectionState.waiting) {
          return Container(color: Theme.of(context).scaffoldBackgroundColor, child: LoadingIndicator());
        }

        if (snapshot.hasError || snapshot.data == null) {
          return _error();
        } else {
          return _announcementScaffold(snapshot.data!);
        }
      },
    );
  }

  Widget _announcementScaffold(AnnouncementViewState announcementViewState) {
    return Scaffold(
      appBar: AppBar(
        title: Text(announcementViewState.toolbarTitle),
      ),
      body: RefreshIndicator(
          onRefresh: () async {
            setState(() {
              _announcementFuture = _loadAnnouncement(context, forceRefresh: true);
            });
            await  _announcementFuture?.catchError((_) { return Future.value(null); });
          },
          child: _announcementBody(announcementViewState)),
    );
  }

  Widget _announcementBody(AnnouncementViewState announcementViewState) {
    return ListView(
      padding: const EdgeInsets.all(16),
      children: <Widget>[
        Text(
          announcementViewState.announcementTitle,
          style: Theme.of(context).textTheme.headlineMedium,
        ),
        SizedBox(height: 4),
        Text(
          announcementViewState.postedAt.l10nFormat(L10n(context).dateAtTime) ?? '',
          style: Theme.of(context).textTheme.bodySmall,
        ),
        SizedBox(height: 16),
        Divider(),
        HtmlDescriptionTile(html: announcementViewState.announcementMessage),
        Divider(),
        SizedBox(height: 16),
        _attachmentsWidget(context, announcementViewState.attachment),
      ],
    );
  }

  Widget _error() {
    return Container(
        color: Theme.of(context).scaffoldBackgroundColor,
        child: ErrorPandaWidget(L10n(context).errorLoadingAnnouncement, () {
          setState(() {
            _announcementFuture = _loadAnnouncement(context, forceRefresh: true);
          });
        }));
  }

  Widget _attachmentsWidget(BuildContext context, Attachment? attachment) {
    if (attachment == null) return Container();
    return Container(
        height: 108,
        padding: EdgeInsets.only(top: 12),
        child: AttachmentIndicatorWidget(
            attachment: attachment,
            onAttachmentClicked: (attachment) => {_interactor.viewAttachment(context, attachment)}));
  }
}

File: apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsScreen.kt
/*
 * Copyright (C) 2024 - present Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
@file:OptIn(ExperimentalMaterialApi::class)

package com.instructure.parentapp.features.alerts.details

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.Divider
import androidx.compose.material.ExperimentalMaterialApi
import androidx.compose.material.Scaffold
import androidx.compose.material.SnackbarHost
import androidx.compose.material.SnackbarHostState
import androidx.compose.material.SnackbarResult
import androidx.compose.material.Text
import androidx.compose.material.pullrefresh.PullRefreshIndicator
import androidx.compose.material.pullrefresh.pullRefresh
import androidx.compose.material.pullrefresh.rememberPullRefreshState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.instructure.canvasapi2.models.Attachment
import com.instructure.canvasapi2.utils.ContextKeeper
import com.instructure.canvasapi2.utils.DateHelper
import com.instructure.pandares.R
import com.instructure.pandautils.compose.CanvasTheme
import com.instructure.pandautils.compose.composables.CanvasThemedAppBar
import com.instructure.pandautils.compose.composables.ComposeCanvasWebViewWrapper
import com.instructure.pandautils.compose.composables.ErrorContent
import com.instructure.pandautils.compose.composables.Loading
import com.instructure.pandautils.features.inbox.utils.AttachmentCard
import com.instructure.pandautils.features.inbox.utils.AttachmentCardItem
import com.instructure.pandautils.features.inbox.utils.AttachmentStatus
import com.instructure.pandautils.views.CanvasWebView
import com.jakewharton.threetenabp.AndroidThreeTen
import java.util.Date

@Composable
fun AnnouncementDetailsScreen(
    uiState: AnnouncementDetailsUiState,
    actionHandler: (AnnouncementDetailsAction) -> Unit,
    navigationActionClick: () -> Unit,
    applyOnWebView: (CanvasWebView.() -> Unit),
    modifier: Modifier = Modifier
) {
    CanvasTheme {
        val snackbarHostState = remember { SnackbarHostState() }
        val errorMessage = stringResource(id = R.string.errorLoadingAnnouncement)
        LaunchedEffect(uiState.showErrorSnack) {
            if (uiState.showErrorSnack) {
                val result = snackbarHostState.showSnackbar(errorMessage)
                if (result == SnackbarResult.Dismissed) {
                    actionHandler(AnnouncementDetailsAction.SnackbarDismissed)
                }
            }
        }

        Scaffold(
            backgroundColor = colorResource(id = R.color.backgroundLightest),
            topBar = {
                CanvasThemedAppBar(
                    title = uiState.pageTitle.orEmpty(),
                    backgroundColor = Color(uiState.studentColor),
                    contentColor = colorResource(id = R.color.textLightest),
                    navigationActionClick = {
                        navigationActionClick()
                    }
                )
            },
            snackbarHost = { SnackbarHost(hostState = snackbarHostState) },
            content = { padding ->
                val pullRefreshState = rememberPullRefreshState(
                    refreshing = uiState.isRefreshing,
                    onRefresh = {
                        actionHandler(AnnouncementDetailsAction.Refresh)
                    }
                )
                Box(modifier = modifier.padding(padding).pullRefresh(pullRefreshState)) {
                    when {
                        uiState.isError -> {
                            ErrorContent(
                                errorMessage = stringResource(id = R.string.errorLoadingAnnouncement),
                                retryClick = {
                                    actionHandler(AnnouncementDetailsAction.Refresh)
                                }, modifier = Modifier.fillMaxSize()
                            )
                        }

                        uiState.isLoading -> {
                            Loading(
                                modifier = Modifier
                                    .fillMaxSize()
                                    .testTag("loading"),
                                color = Color(uiState.studentColor)
                            )
                        }

                        else -> {
                            AnnouncementDetailsSuccessScreen(
                                uiState,
                                actionHandler,
                                applyOnWebView,
                                modifier
                            )
                        }
                    }
                    PullRefreshIndicator(
                        refreshing = uiState.isRefreshing,
                        state = pullRefreshState,
                        modifier = Modifier
                            .align(Alignment.TopCenter)
                            .testTag("pullRefreshIndicator"),
                        contentColor = Color(uiState.studentColor)
                    )
                }
            })
    }
}

@Composable
private fun AnnouncementDetailsSuccessScreen(
    uiState: AnnouncementDetailsUiState,
    actionHandler: (AnnouncementDetailsAction) -> Unit,
    applyOnWebView: (CanvasWebView.() -> Unit),
    modifier: Modifier = Modifier
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
    ) {
        val padding = 16.dp
        Column(
            modifier = modifier.padding(vertical = padding)
        ) {
            Column(modifier = Modifier.padding(horizontal = padding)) {
                Text(
                    text = uiState.announcementTitle.orEmpty(),
                    color = colorResource(id = R.color.textDarkest),
                    fontSize = 24.sp
                )
                Text(
                    text = DateHelper.getDateAtTimeString(
                        LocalContext.current,
                        R.string.alertDateTime,
                        uiState.postedDate
                    ).orEmpty(),
                    modifier = Modifier.padding(top = 4.dp, bottom = 18.dp),
                    color = colorResource(id = R.color.textDarkest),
                    fontSize = 14.sp
                )
                AttachmentsRow(uiState.attachment, actionHandler)
            }
            uiState.message?.let { message ->
                Divider(Modifier.padding(horizontal = padding))
                Text(
                    modifier = Modifier.padding(
                        top = 18.dp, bottom = 6.dp, start = padding, end = padding
                    ),
                    text = stringResource(R.string.description),
                    color = colorResource(id = R.color.textDarkest),
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Medium
                )
                ComposeCanvasWebViewWrapper(
                    modifier = Modifier.padding(horizontal = 6.dp),
                    html = message,
                    applyOnWebView = applyOnWebView
                )
            }
        }
    }
}

@Composable
private fun AttachmentsRow(
    attachment: Attachment?,
    actionHandler: (AnnouncementDetailsAction) -> Unit
) {
    attachment?.let {
        AttachmentCard(
            AttachmentCardItem(
                attachment = it,
                status = AttachmentStatus.UPLOADED,
                readOnly = true
            ),
            onSelect = { actionHandler(AnnouncementDetailsAction.OpenAttachment(it)) },
            onRemove = {}
        )
        Spacer(modifier = Modifier.height(12.dp))
    }
}

@Preview
@Composable
private fun AnnouncementDetailsPreview() {
    ContextKeeper.appContext = LocalContext.current
    AndroidThreeTen.init(LocalContext.current)
    AnnouncementDetailsScreen(
        uiState = AnnouncementDetailsUiState(
            pageTitle = "Course Name",
            announcementTitle = "Announcement Title",
            postedDate = Date(),
            message = "",
            attachment = null
        ),
        actionHandler = {},
        navigationActionClick = {},
        applyOnWebView = {}
    )
}

@Preview
@Composable
private fun AnnouncementDetailsAttachmentPreview() {
    ContextKeeper.appContext = LocalContext.current
    AndroidThreeTen.init(LocalContext.current)
    AnnouncementDetailsScreen(
        uiState = AnnouncementDetailsUiState(
            pageTitle = "Course Name",
            announcementTitle = "Announcement Title",
            postedDate = Date(),
            message = "",
            attachment = Attachment(
                id = 1,
                filename = "Attached_document",
                contentType = "pdf"
            )
        ),
        actionHandler = {},
        navigationActionClick = {},
        applyOnWebView = {}
    )
}

@Preview
@Composable
private fun AnnouncementDetailsErrorPreview() {
    AnnouncementDetailsScreen(
        uiState = AnnouncementDetailsUiState(
            isError = true
        ),
        actionHandler = {},
        navigationActionClick = {},
        applyOnWebView = {}
    )
}

@Preview
@Composable
private fun AnnouncementDetailsLoadingPreview() {
    AnnouncementDetailsScreen(
        uiState = AnnouncementDetailsUiState(
            isLoading = true
        ),
        actionHandler = {},
        navigationActionClick = {},
        applyOnWebView = {}
    )
}
